File: src/model/Inventory.java
package model;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Random;

import model.Item.ITEM_TYPE;

import core.Logger;

/**
 * Inventory with an list that holds all the arrayLists of items in the inventory.
 */
public class Inventory implements Serializable {
	private final List<PriorityQueue<Item>> slots;
	
	private final int MAX_SIZE;
	
	private final double MAX_WEIGHT;
	
	private int currentSize;

	/**
	 * Inventory starts out empty.
	 * @param maxSize the max size of the inventory
	 * @param maxWeight the max weight of the inventory
	 */
	public Inventory(int maxSize, double maxWeight) {
		this.MAX_SIZE = maxSize;
		this.MAX_WEIGHT = maxWeight;
		this.slots = new ArrayList<PriorityQueue<Item>>(Item.ITEM_TYPE.values().length);
		for(int i = 0; i < Item.ITEM_TYPE.values().length; i++) {
			slots.add(new PriorityQueue<Item>());
		}
		this.currentSize = 0;
	}
	
	/**
	 * Returns the max number of items we can hold.
	 * @return Max number of items to hold
	 */
	public int getMaxSize() {
		return MAX_SIZE;
	}
	
	/**
	 * Returns the slots in inventory with items.
	 * @return The slots in the inventory.
	 */
	public List<Item.ITEM_TYPE> getPopulatedSlots() {
		final List<Item.ITEM_TYPE> popSlots = new ArrayList<Item.ITEM_TYPE>();
		for(Item.ITEM_TYPE itemType : Item.ITEM_TYPE.values()) {
			if (getNumberOf(itemType) != 0) {
				popSlots.add(itemType);
			}
		}
		return popSlots;
	}
	
	/**
	 * Returns the current size of the inventory (number of item queues with contents)
	 * @return The size of the inventory
	 */
	public int getCurrentSize() {
		return currentSize;
	}
	
	/**
	 * Calculates and returns the weight of the inventory.
	 * @return The weight of the inventory.
	 */
	public double getWeight() {
		double weight = 0;
		for(PriorityQueue<Item> slot : slots) {
			for(Item item : slot) {
				weight += item.getWeight();
			}
		}
		return weight;
	}

	/**
	 * Determines if the inventory can handle the addition
	 * @param itemType The item type
	 * @param numberOf The number of items to test with
	 * @return True if successful
	 */
	public boolean canGetItems(Item.ITEM_TYPE itemType, int numberOf) {
		if(numberOf == 0) {
			return false;
		}
		
		final double weight = itemType.getWeight() * numberOf;
		if(getWeight() + weight > MAX_WEIGHT) {
			Logger.log("Not enough weight capacity", Logger.Level.DEBUG);
			return false;
		} else if (currentSize == MAX_SIZE && getNumberOf(itemType) == 0) {
			Logger.log("Not enough slots open", Logger.Level.DEBUG);
			return false;
		} else {
			return true;
		}
	}
	
	/**
	 * Adds the item to the inventory.
	 * @param itemsToAdd The items to add
	 */
	public void addItemsToInventory(List<Item> itemsToAdd) {
		if(itemsToAdd.size() == 0) {
			return;
		}
		final Item.ITEM_TYPE itemType = itemsToAdd.get(0).getType();
		if(canGetItems(itemType, itemsToAdd.size())) {
			for(Item item : itemsToAdd) {
				slots.get(itemType.ordinal()).add(item);
				Logger.log(item.getName() + " added.", Logger.Level.DEBUG);
			}
			currentSize = 0;
			for(PriorityQueue<Item> slot : slots) {
				if (slot.size() > 0) {
					currentSize += 1;
				}
			}
			return;
		} else {
			Logger.log("Add item failed", Logger.Level.DEBUG);
			return;
		}
	}
		
	public void addItemToInventory(Item item) {
		final List<Item> itemToAdd = new ArrayList<Item>();
		itemToAdd.add(item);
		addItemsToInventory(itemToAdd);
	}

	/**
	 * Removes the item from the inventory and gives it back as an arrayList
	 * @param itemType The type of item to remove
	 * @param quantity The number of the item to remvove
	 * @return The removed items.
	 */
	public List<Item> removeItemFromInventory(Item.ITEM_TYPE itemType, int quantity) {
		List<Item> removedItems = new ArrayList<Item>();
		final int itemIndex = itemType.ordinal();
		if(slots.get(itemIndex).size() < quantity) {
			Logger.log("Not enough items to remove", Logger.Level.DEBUG);
			removedItems = null;
		} else {
			for(int i = 0; i < quantity; i++) {
				removedItems.add(slots.get(itemIndex).poll());
			}
			Logger.log("Items removed successfully", Logger.Level.DEBUG);
		}
		
		currentSize = 0;
		for(PriorityQueue<Item> slot : slots) {
			if (slot.size() > 0) {
				currentSize += 1;
			}
		}
		
		return removedItems;
	}
	
	/**
	 * Checks to see if the current inventory is full.
	 * @return True if the inventory is full, false otherwise.
	 */
	public boolean isFull() {
		return (slots.size() == MAX_SIZE);
	}
	
	/**
	 * Checks to see if the current inventory is empty.
	 * @return True if the inventory is empty, false otherwise.
	 */
	public boolean isEmpty() {
		return (getCurrentSize() == 0);
	}
	
	/**
	 * Returns the number of a specific type of item in the inventory
	 * @param itemType The item type queried
	 * @return The number of that item type.
	 */
	public int getNumberOf(Item.ITEM_TYPE itemType) {
		return slots.get(itemType.ordinal()).size();
	}
	
	/**
	 * Returns the number of any item type in this inventory.
	 * @return The number of any item types that is in this inventory
	 */
	public int getNumberOfItems() {
		int size = 0;
		
		for(PriorityQueue<Item> slot : slots) {
			size += slot.size();
		}
		
		return size;
	}
	
	/**
	 * Checks the item type
	 * @param itemType The item to check
	 * @return The condition of the worst item in the list
	 */
	public Condition getConditionOf(Item.ITEM_TYPE itemType) {
		return getNumberOf(itemType) != 0 ? 
				slots.get(itemType.ordinal()).peek().getStatus() :
					null;
	}
	
	/**
	 * Clear all the items in each slot.
	 */
	public void clear() {
		for(PriorityQueue<Item> slot : slots) {
			slot.clear();
		}
		
		currentSize = 0;
	}
	
	/**
	 * Add random items to this inventory.
	 */
	public void addRandomItems() {
		Random random = new Random();
		
		int numberOfItemsToAdd = random.nextInt(getMaxSize()) + 2;
		
		for (int i = 0; i < numberOfItemsToAdd; i++) {
			Item item;
			int attempts = 0;
			do {
				int randomItem = random.nextInt(ITEM_TYPE.values().length);
				item = new Item(ITEM_TYPE.values()[randomItem]);
				
				attempts++;
			} while(!canGetItems(item.getType(), 1) && attempts < ITEM_TYPE.values().length);
			
			item.decreaseStatus(random.nextInt(101));
			if (!item.getType().isAnimal() && 
				item.getType() != ITEM_TYPE.STRANGEMEAT && 
				item.getType() != ITEM_TYPE.WAGON &&
				item.getType() != ITEM_TYPE.SONIC
				)
			addItemToInventory(item);
		}
	}
	
	/**
	 * ToString method for debugging and whatnot
	 * @return String representation of the inventory
	 */
	public String toString() {
		final List<Item.ITEM_TYPE> popSlots = getPopulatedSlots();
		final StringBuffer str = new StringBuffer("Size: " + popSlots.size() + ". ");
		for(Item.ITEM_TYPE itemType : popSlots) {
			str.append(" # of " + itemType.getName() + "s: " + getNumberOf(itemType));
		}
				
		return str.toString();
	}

	public double getMaxWeight() {
		return MAX_WEIGHT;
	}
}

File: web/src/model/Inventory.ts
import { Item } from "./Item";
import { ItemType, isItemTypeAnimal } from "./ItemType";
import { Condition } from "./Condition";

export class Inventory {
  private readonly slots: Item[][];
  private readonly MAX_SIZE: number;
  private readonly MAX_WEIGHT: number;
  private currentSize: number;

  constructor(maxSize: number, maxWeight: number) {
    this.MAX_SIZE = maxSize;
    this.MAX_WEIGHT = maxWeight;
    this.currentSize = 0;

    this.slots = [];
    const itemTypes = Object.values(ItemType);
    for (let i = 0; i < itemTypes.length; i++) {
      this.slots.push([]);
    }
  }

  getMaxSize(): number {
    return this.MAX_SIZE;
  }

  getPopulatedSlots(): ItemType[] {
    const popSlots: ItemType[] = [];
    const itemTypes = Object.values(ItemType);
    for (const itemType of itemTypes) {
      if (this.getNumberOf(itemType) !== 0) {
        popSlots.push(itemType);
      }
    }
    return popSlots;
  }

  getCurrentSize(): number {
    return this.currentSize;
  }

  getWeight(): number {
    let weight = 0;
    for (const slot of this.slots) {
      for (const item of slot) {
        weight += item.getWeight();
      }
    }
    return weight;
  }

  canGetItems(itemType: ItemType, numberOf: number): boolean {
    if (numberOf <= 0) {
      return false;
    }

    if (isItemTypeAnimal(itemType)) {
      return true;
    }

    const weight = itemType ? this.getItemTypeWeight(itemType) * numberOf : 0;
    if (this.getWeight() + weight > this.MAX_WEIGHT) {
      console.debug("Not enough weight capacity");
      return false;
    } else if (
      this.currentSize === this.MAX_SIZE &&
      this.getNumberOf(itemType) === 0
    ) {
      console.debug("Not enough slots open");
      return false;
    }

    return true;
  }

  private getItemTypeWeight(itemType: ItemType): number {
    const itemTypes = Object.values(ItemType);
    const ordinal = itemTypes.indexOf(itemType);
    if (ordinal === -1 || this.slots[ordinal].length === 0) {
      return 0;
    }
    return this.slots[ordinal][0].getWeight();
  }

  addItemsToInventory(itemsToAdd: Item[]): void {
    if (itemsToAdd.length === 0) {
      return;
    }

    const itemType = itemsToAdd[0].getType();
    if (!this.canGetItems(itemType, itemsToAdd.length)) {
      console.debug("Add item failed");
      return;
    }

    const itemTypes = Object.values(ItemType);
    const ordinal = itemTypes.indexOf(itemType);
    for (const item of itemsToAdd) {
      this.slots[ordinal].push(item);
      console.debug(`${item.getName()} added.`);
    }

    this.slots[ordinal].sort((a, b) => a.compareTo(b));

    this.recalculateCurrentSize();
  }

  addItemToInventory(item: Item): void {
    this.addItemsToInventory([item]);
  }

  removeItemFromInventory(
    itemType: ItemType,
    quantity: number,
  ): Item[] | null {
    const itemTypes = Object.values(ItemType);
    const itemIndex = itemTypes.indexOf(itemType);

    if (this.slots[itemIndex].length < quantity) {
      console.debug("Not enough items to remove");
      return null;
    }

    const removedItems: Item[] = [];
    for (let i = 0; i < quantity; i++) {
      const item = this.slots[itemIndex].shift();
      if (item) {
        removedItems.push(item);
      }
    }
    console.debug("Items removed successfully");

    this.recalculateCurrentSize();
    return removedItems;
  }

  private recalculateCurrentSize(): void {
    this.currentSize = 0;
    for (const slot of this.slots) {
      if (slot.length > 0) {
        this.currentSize += 1;
      }
    }
  }

  isFull(): boolean {
    return this.slots.length === this.MAX_SIZE;
  }

  isEmpty(): boolean {
    return this.getCurrentSize() === 0;
  }

  getNumberOf(itemType: ItemType): number {
    const itemTypes = Object.values(ItemType);
    const ordinal = itemTypes.indexOf(itemType);
    return this.slots[ordinal].length;
  }

  getNumberOfItems(): number {
    let size = 0;
    for (const slot of this.slots) {
      size += slot.length;
    }
    return size;
  }

  getConditionOf(itemType: ItemType): Condition | null {
    const count = this.getNumberOf(itemType);
    if (count === 0) {
      return null;
    }

    const itemTypes = Object.values(ItemType);
    const ordinal = itemTypes.indexOf(itemType);
    return this.slots[ordinal][0].getStatus();
  }

  clear(): void {
    for (const slot of this.slots) {
      slot.length = 0;
    }
    this.currentSize = 0;
  }

  addRandomItems(): void {
    const numberOfItemsToAdd = Math.floor(Math.random() * this.getMaxSize()) + 2;

    const itemTypes = Object.values(ItemType);
    for (let i = 0; i < numberOfItemsToAdd; i++) {
      let item: Item;
      let attempts = 0;

      do {
        const randomItem = Math.floor(Math.random() * itemTypes.length);
        item = new Item(itemTypes[randomItem]);
        attempts++;
      } while (!this.canGetItems(item.getType(), 1) && attempts < itemTypes.length);

      item.decreaseStatus(Math.floor(Math.random() * 101));

      const itemType = item.getType();
      if (
        !isItemTypeAnimal(itemType) &&
        itemType !== ItemType.STRANGEMEAT &&
        itemType !== ItemType.WAGON &&
        itemType !== ItemType.SONIC &&
        itemType !== ItemType.MAP
      ) {
        this.addItemToInventory(item);
      }
    }
  }

  toString(): string {
    const popSlots = this.getPopulatedSlots();
    let str = `Size: ${popSlots.length}. `;
    for (const itemType of popSlots) {
      const itemTypes = Object.values(ItemType);
      const ordinal = itemTypes.indexOf(itemType);
      const name = this.slots[ordinal][0]?.getName() || "";
      str += ` # of ${name}s: ${this.getNumberOf(itemType)}`;
    }
    return str;
  }

  getMaxWeight(): number {
    return this.MAX_WEIGHT;
  }
}
